File: src/test/java/CommandTest.java
import com.google.common.collect.Lists;
import com.networkedassets.plugins.managemytree.*;
import com.networkedassets.plugins.managemytree.commands.*;
import org.codehaus.jackson.map.ObjectMapper;
import org.junit.Test;

import java.io.IOException;
import java.util.Collections;

import static org.junit.Assert.assertEquals;

public class CommandTest {
    private final ObjectMapper om = new ObjectMapper();

    @Test
    public void testSerializeCommand() throws IOException {
        final AddPage ap = new AddPage("a", "j1", "2");
        ap.setConfluenceId(2L);
        final String aps = om.writeValueAsString(ap);
        final Command apd = om.readValue(aps, Command.class);
        assertEquals(ap, apd);

        final RemovePage rp = new RemovePage("0");
        rp.getRemovedPages().add(new OriginalPage(0, new Location(1, 2)));
        rp.setName("foo");
        final String rps = om.writeValueAsString(rp);
        final Command rpd = om.readValue(rps, Command.class);
        assertEquals(rp, rpd);

        final MovePage mp = new MovePage("1", null, 0);
        mp.setMovedPage(new OriginalPage(1, new Location(2, 3)));
        mp.setName("foo");
        final String mps = om.writeValueAsString(mp);
        final Command mpd = om.readValue(mps, Command.class);
        assertEquals(mp, mpd);

        final RenamePage rnp = new RenamePage("42", "czterdzieści dwa");
        rnp.setOldName("lel");
        rnp.setRenamedPageId(42L);
        final String rnps = om.writeValueAsString(rnp);
        final Command rnpd = om.readValue(rnps, Command.class);
        assertEquals(rnp, rnpd);

        final InsertTemplate inp = new InsertTemplate(
                "foo",
                new TemplateId.FromBlueprint("bar"),
                Collections.singletonMap(2, "b")
        );
        inp.getInsertedPages().add(2L);
        final String inps = om.writeValueAsString(inp);
        final Command inpd = om.readValue(inps, Command.class);
        assertEquals(inp, inpd);

        final InsertTemplate inp2 = new InsertTemplate(
                "foo",
                new TemplateId.Custom(42),
                Collections.singletonMap(1, "a")
        );
        inp2.getInsertedPages().add(3L);
        final String inps2 = om.writeValueAsString(inp2);
        final Command inpd2 = om.readValue(inps2, Command.class);
        assertEquals(inp2, inpd2);
    }

    @Test
    public void testSerializeOutline() throws IOException {
        final CustomOutline o = new CustomOutline("foo", Lists.newArrayList(), null);
        final String s = "{\"title\": \"foo\", \"children\": []}";
        CustomOutline customOutline = om.readValue(s, CustomOutline.class);
        assertEquals(customOutline, o);

        final CustomTemplate t = new CustomTemplate("bar", o);
        CustomTemplate customTemplate = om.readValue(om.writeValueAsString(t), CustomTemplate.class);
        assertEquals(customTemplate, t);
    }
}

File: src/test/java/CommandTest.kt

import com.google.common.collect.Lists
import com.networkedassets.plugins.managemytree.*
import com.networkedassets.plugins.managemytree.commands.*
import org.codehaus.jackson.map.ObjectMapper
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

class CommandTest {
    private val om = ObjectMapper()

    @Test
    fun testSerializeCommand() {
        val ap = AddPage("a", "j1", "2")
        ap.confluenceId = 2L
        val aps = om.writeValueAsString(ap)
        val apd = om.readValue(aps, Command::class.java)
        assertEquals(ap, apd)

        val rp = RemovePage("0")
        rp.removedPages.add(OriginalPage(0, Location(1, 2)))
        rp.name = "foo"
        val rps = om.writeValueAsString(rp)
        val rpd = om.readValue(rps, Command::class.java)
        assertEquals(rp, rpd)

        val mp = MovePage("1", null, 0)
        mp.movedPage = OriginalPage(1, Location(2, 3))
        mp.name = "foo"
        val mps = om.writeValueAsString(mp)
        val mpd = om.readValue(mps, Command::class.java)
        assertEquals(mp, mpd)

        val rnp = RenamePage("42", "czterdzieści dwa")
        rnp.oldName = "lel"
        rnp.renamedPageId = 42L
        val rnps = om.writeValueAsString(rnp)
        val rnpd = om.readValue(rnps, Command::class.java)
        assertEquals(rnp, rnpd)

        val inp = InsertTemplate(
                "foo",
                TemplateId.FromBlueprint("bar"),
                Collections.singletonMap("2", "b"))
        inp.insertedPages.add(2L)
        val inps = om.writeValueAsString(inp)
        val inpd = om.readValue(inps, Command::class.java)
        assertEquals(inp, inpd)

        val inp2 = InsertTemplate(
                "foo",
                TemplateId.Custom(42),
                Collections.singletonMap("2", "a"))
        inp2.insertedPages.add(3L)
        val inps2 = om.writeValueAsString(inp2)
        val inpd2 = om.readValue(inps2, Command::class.java)
        assertEquals(inp2, inpd2)
    }

    @Test
    fun testSerializeOutline() {
        val o = CustomOutline("foo", Lists.newArrayList<CustomOutline>(), null)
        //language=JSON
        val s = """{"title": "foo", "children": []}"""
        val customOutline = om.readValue(s, CustomOutline::class.java)
        assertEquals(customOutline, o)

        val t = CustomTemplate("bar", listOf(o))
        val customTemplate = om.readValue(om.writeValueAsString(t), CustomTemplate::class.java)
        assertEquals(customTemplate, t)
    }

    @Test
    fun testSerializeKotlinObjectLiterals() {
        val s = om.writeValueAsString(object { val name = "foo"; val id = 2; })
        assertEquals(s, """{"name":"foo","id":2}""")
    }

    @Test
    fun testDeserializeInsertTemplate() {
        //language=JSON
        val s = """
            {
                "commandType": "insertTemplate",
                "templateId": {
                    "templateType": "custom",
                    "customOutlineId": 1
                },
                "parentId": "98310",
                "newPageJstreeIds": {
                    "1": "j1_11",
                    "2": "j1_12"
                },
                "name": "test"
            }
        """;

        val cmds: Command = om.deserialize(s);
    }

}

inline fun <reified T: Any> ObjectMapper.deserialize(s: String): T =
        this.readValue(s, T::class.java)
